src/megaapi_impl.cpp

```diff
@@ -11380,42 +11323,126 @@ bool MegaApiImpl::processTree(Node* node, TreeProcessor* processor, bool recursi
-    for (node_list::iterator it = node->children.begin(); it != node->children.end()
-         && !(cancelToken && cancelToken->isCancelled()); )
+    
+    MegaNodeList *nodeList = nullptr;
+    if (n)
     {
-        processTree(*it++, &searchProcessor, recursive, cancelToken);
+        // if node is provided, it will be the parent node of the tree to explore
+        Node *node = client->nodebyhandle(n->getHandle());
+        if (!node)
+        {
+            return new MegaNodeListPrivate();
+        }
+
+        // searchString and nodeType (if provided), are considered in search
+        SearchTreeProcessor searchProcessor(client, searchString, static_cast<nodefiletype_t>(type));
+        for (node_list::iterator it = node->children.begin(); it != node->children.end()
+             && !(cancelToken && cancelToken->isCancelled()); )
+        {
+            processTree(*it++, &searchProcessor, recursive, cancelToken);
+        }
+
+        vector<Node *>& vNodes = searchProcessor.getResults();
+        sortByComparatorFunction(vNodes, order, *client);
+        nodeList = new MegaNodeListPrivate(vNodes.data(), int(vNodes.size()));
     }
+    else
+    {
+        node_vector result;
+        Node *node;
+
+        // Target parameter is only considered if node is not provided
+        if (target < MegaApi::TARGET_INSHARE || target > MegaApi::TARGET_ALL)
+        {
+            return new MegaNodeListPrivate();
+        }
+
+        if (target == MegaApi::TARGET_ROOTNODE || target == MegaApi::TARGET_ALL)
+        {
+            // Search on rootnode (cloud, excludes Inbox and Rubbish)
+            node = client->nodebyhandle(client->rootnodes[0]);
 
-    vector<Node *>& vNodes = searchProcessor.getResults();
-    sortByComparatorFunction(vNodes, order, *client);
+            SearchTreeProcessor searchProcessor(client, searchString, static_cast<nodefiletype_t>(type));
+            processTree(node, &searchProcessor, recursive, cancelToken);
+            node_vector& vNodes = searchProcessor.getResults();
+
+            result.insert(result.end(), vNodes.begin(), vNodes.end());
+        }
+
+        if (target == MegaApi::TARGET_INSHARE || target == MegaApi::TARGET_ALL)
+        {
+            // Search on inshares
+            unique_ptr<MegaShareList> shares(getInSharesList(MegaApi::ORDER_NONE));
+            for (int i = 0; i < shares->size() && !(cancelToken && cancelToken->isCancelled()); i++)
+            {
+                node = client->nodebyhandle(shares->get(i)->getNodeHandle());
+
+                SearchTreeProcessor searchProcessor(client, searchString, static_cast<nodefiletype_t>(type));
+                processTree(node, &searchProcessor, recursive, cancelToken);
+                vector<Node *>& vNodes  = searchProcessor.getResults();
+
+                result.insert(result.end(), vNodes.begin(), vNodes.end());
+            }
+        }
+
+        if (target == MegaApi::TARGET_OUTSHARE || target == MegaApi::TARGET_ALL)
```

Comment:
I think with this check is enough 
```suggestion
        if (target == MegaApi::TARGET_OUTSHARE)
```
Out shares are inside our cloud and they are search at `if (target == MegaApi::TARGET_ROOTNODE || target == MegaApi::TARGET_ALL)`
I think public link can be done at same way

Reply:
You are right, I've adjusted in commit:  2db8ad48b9d51af1d41025f827bdcd8a4e84ec2e